include/wil/com_apartment_variable.h

```diff
@@ -0,0 +1,230 @@
+
+        unsigned long long apartmentId{};
+        FAIL_FAST_IF_FAILED(RoGetApartmentIdentifier(&apartmentId));
+
+        const auto variableKey = std::addressof(createFn);
+
+        auto lock = winrt::slim_lock_guard(details::s_lock);
+
+        auto storage = details::s_apartmentStorage.find(apartmentId);
+        if (storage != details::s_apartmentStorage.end())
+        {
+            auto& variables = storage->second;
+            auto variable = variables.find(variableKey);
+            if (variable != variables.end())
+            {
+                return std::any_cast<std::invoke_result_t<F>>(variable->second);
+            }
+            else
+            {
+                auto result = createFn();
+                variables.insert({ variableKey, result });
+#ifdef _DEBUG
+                variable = variables.find(variableKey);
+                assert(variable != variables.end());
+                std::ignore = std::any_cast<std::invoke_result_t<F>>(variable->second);
+#endif
+                return result;
+            }
+        }
+        else
+        {
+            struct ApartmentObserver : public winrt::implements<ApartmentObserver, IApartmentShutdown>
+            {
+                void STDMETHODCALLTYPE OnUninitialize(unsigned long long apartmentId) noexcept override
+                {
+                    // This code runs at apartment rundown so be careful to avoid deadlocks by
+                    // extracting the variables under the lock then release them outside.
+                    auto variables = [apartmentId]()
+                    {
+                        auto lock = winrt::slim_lock_guard(details::s_lock);
+                        return details::s_apartmentStorage.extract(apartmentId);
+                    }();
+                }
+            };
+
+            auto observer = winrt::make<ApartmentObserver>();
+            unsigned long long id{};
+            APARTMENT_SHUTDOWN_REGISTRATION_COOKIE cookie{}; // leaked, COM run down releases all instances
+            winrt::check_hresult(RoRegisterForApartmentShutdown(observer.get(), &id, &cookie));
+            FAIL_FAST_IF(apartmentId != id);
+
+            auto result = createFn();
+            details::s_apartmentStorage.insert({ apartmentId,
+                decltype(details::s_apartmentStorage)::value_type::second_type { { variableKey, result } } });
+#ifdef _DEBUG
+            auto a = details::s_apartmentStorage.find(id);
+            assert(a != details::s_apartmentStorage.end());
+
+            auto& variables = a->second;
+            auto variable = variables.find(variableKey);
+            assert(variable != variables.end());
+            std::ignore = std::any_cast<std::invoke_result_t<F>>(variable->second);
+#endif
+            return result;
+        }
+    }
+
+    // Remove the variable including its entry in the map. This has no
+    // effect if the variable was not stored based on a call to get_for_current_com_apartment

```

Comment:
I don't see why this restriction is necessary. It means that people are going to have a dummy get before the reset.

```cpp
auto DefaultCurrentWidget() { return std::_weak_ptr<Widget>(); }

void SetCurrentWidgetForThread(std::shared_ptr<Widget> const& widget)
{
    // dummy get_() call to populate the entry so we can reset it
    std::ignore = wil:get_for_current_com_apartment(DefaultCurrentWidget);

    wil::reset_for_current_com_apartment(DefaultCurrentWidget, widget);
}
```

Reply:
did you mean to comment on the strict version? this was the open design question, I took your feedback, assuming it was on the strict version to choose the non-strict implementation (this one). I also chose the strict version with `newValue`. 

let me know if I understood you correclty.